crates/biome_js_analyze/src/lint/nursery/no_template_curly_in_string.rs

```diff
@@ -0,0 +1,95 @@
+use biome_analyze::{
+    context::RuleContext, declare_lint_rule, Ast, Rule, RuleDiagnostic, RuleSource,
+};
+use biome_console::markup;
+use biome_js_syntax::JsStringLiteralExpression;
+use biome_rowan::AstNode;
+
+declare_lint_rule! {
+    /// Disallow template literal placeholder syntax in regular strings.
+    ///
+    /// ECMAScript 6 allows programmers to create strings containing variable or expressions using template literals,
+    /// instead of string concatenation, by writing expressions like ${variable} between two backtick quotes (`).
+    /// It can be easy to use the wrong quotes when wanting to use template literals, by writing "${variable}",
+    /// and end up with the literal value "${variable}" instead of a string containing the value of the injected expressions.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// const a = "Hello ${name}!";
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// const a = 'Hello ${name}!';
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// const a = "Time: ${12 * 60 * 60 * 1000}";
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// const a = `Hello ${name}!`;
+    /// const a = `Time: ${12 * 60 * 60 * 1000}`;
+    ///
+    /// const a = templateFunction`Hello ${name}`;
+    /// ```
+    ///
+    pub NoTemplateCurlyInString {
+        version: "next",
+        name: "noTemplateCurlyInString",
+        language: "js",
+        sources: &[RuleSource::Eslint("no-template-curly-in-string")],
+        recommended: false,
+    }
+}
+
+impl Rule for NoTemplateCurlyInString {
+    type Query = Ast<JsStringLiteralExpression>;
+    type State = ();
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+        let token = node.value_token().ok()?;
+        let text = token.text();
+
+        let bytes = text.as_bytes();
+        let mut byte_iter = bytes.iter();
+
+        while let Some(&byte) = byte_iter.next() {
+            if byte == b'$' {
+                if let Some(&next_byte) = byte_iter.next() {
+                    if next_byte == b'{' {
+                        for &inner_byte in byte_iter.by_ref() {
+                            if inner_byte == b'}' {
+                                return Some(());
+                            }
+                        }
+                    }
+                }
+            }
+        }
```

Comment:
I completely forget to say:
- you can directly use `text.bytes()` to get a iterator of bytes.
  This has the advantage of getting bytes instead of reference to bytes.
- We could report the precise range of the placeholder. This will help users.
  To do that, we could use `enumerate()` and return a range of `u32` as state.
  In the `diagnostic` method, we could build a `TextRange` from the range of `u32` and use it to report the issue.
```suggestion
        let mut byte_iter = text.bytes().enumerate();

        while let Some((i, byte)) = byte_iter.next() {
            if byte == b'$' {
                if let Some((_, b'{')) = byte_iter.next() {
                    while  let Some((j, byte)) = byte_iter.next() {
                        if inner_byte == b'}' {
                            return Some((i as u32, (j + 1) as u32));
                        }
                    }
                }
            }
        }
```

Reply:
updated both the code and `range()` here, the `invalid.js.snapshot` can show this.